### DISCUSSION TITLE
CefSharp.Common.NETCore some dlls target framework is not .netCore, it is still .NetFramwork

### DISCUSSION THREAD

**Discussion by thudugala** [cefsharp/CefSharp#3689] [General]
CefSharp.Common.NETCore some dlls target framework is not .netCore, it is still .NetFramwork

![image](https://user-images.githubusercontent.com/4112014/125871350-54f5eb20-8420-4a2a-9daa-93aa45cd67b5.png)

![image](https://user-images.githubusercontent.com/4112014/125871423-81ef9068-02f7-403b-a24c-606076cf7639.png)

Due to this I'm getting below error

Could not load file or assembly 'CefSharp.Core.Runtime, Version=91.1.211.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format.

when .Net 4.8 is turned off in the computer

![image](https://user-images.githubusercontent.com/4112014/125871499-5720dcb9-266e-4695-a57a-ac931e1ea2d8.png)

Is there any alternative to fix this ?

**Comment by amaitland** [comment]
> CefSharp.Common.NETCore some dlls target framework is not .netCore, it is still .NetFramwork

The dlls you've highlighted are mixed mode C++/CLI dlls. They are built with a target framework of. Net Core 3.1.

https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.Core.Runtime/CefSharp.Core.Runtime.netcore.vcxproj#L31
https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.BrowserSubprocess.Core/CefSharp.BrowserSubprocess.Core.netcore.vcxproj#L31

What tool are you using to inspect the assemblies? Is it the most up to date version? 

They are not standard CLI assemblies. 

> when .Net 4.8 is turned off in the computer

What OS? Is Visual C++ 2019 installed on the target machine? What version of. Net are you targeting? What version was installed on the target machine? 

https://github.com/cefsharp/CefSharp/blob/cefsharp/91/NuGet/PackageReference/Readme.txt#L22

> Could not load file or assembly 'CefSharp.Core.Runtime, Version=91.1.211.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format.

The most common cause of this error is that Visual C++ 2019 is missing from the target machine.

**Reply by amaitland** [reply]
I should also point out that even if they were targeting. Net 4.0 it wouldn't require your target machine to have. Net 4.0 installed. The .Net Core 3.1 and .Net 5.0 runtimes are perfectly capable of loading a .Net 4.0 assembly if all the required dependencies are present. The runtimes are independent and backward compatible.

**Reply by amaitland** [reply]
Using `dotPeek` version `2021.1.4` downloaded from https://www.jetbrains.com/decompiler/download/#section=portable

If I download https://www.nuget.org/api/v2/package/CefSharp.Common.NETCore/91.1.211 and unzip the `nupkg` file then load the `win-x64` and `win-x86` versions of `CefSharp.Core.Runtime.dll` in `dotPeek` then I get the expected `.Net Core 3.1`.

<img width="364" alt="commonnetcore" src="https://user-images.githubusercontent.com/307872/125874715-568cc792-c2bb-4087-a12c-89942935e259.PNG">

Whatever software you are using is likely too old to correctly identify the dlls.

**Reply by thudugala** [reply]
"cefsharp.common.netcore.91.1.211\runtimes\win-x86\lib\netcoreapp3.1\CefSharp.Core.Runtime.dll"
"cefsharp.common.netcore.91.1.211\runtimes\win-x64\lib\netcoreapp3.1\CefSharp.Core.Runtime.dll"

are reporting as .Net Framwork 4.0

"cefsharp.common.netcore.91.1.211\ref\netcoreapp3.1\CefSharp.Core.Runtime.dll"

is reporting as netcoreapp3.1

my application is building as win-x86, CefSharp.Core.Runtime.dll in bin folder is .Net Framwork 4.0, I assume that is comming from "cefsharp.common.netcore.91.1.211\runtimes\win-x86\lib\netcoreapp3.1\CefSharp.Core.Runtime.dll"

![image](https://user-images.githubusercontent.com/4112014/125876327-5a3a755f-1e97-445b-a37a-2bb8bf3cbcd3.png)

**Reply by amaitland** [reply]
> are reporting as .Net Framwork 4.0

What software are you using to inspect the dlls? Whatever software you are using is likely too old to correctly identify the dlls.

They are both .Net Core 3.1.

Please download the latest version of `dotPeek` and check with that.

> "cefsharp.common.netcore.91.1.211\ref\netcoreapp3.1\CefSharp.Core.Runtime.dll"
> is reporting as netcoreapp3.1

The `ref` assembly is a managed assembly, unlike the others which are mixed mode `C++/CLI`. The ref assembly is used at compile time.

**Reply by thudugala** [reply]
I'm using the same version

![image](https://user-images.githubusercontent.com/4112014/125877698-50666921-4698-4f5d-84b5-e8b120141925.png)

**Reply by amaitland** [reply]
I don't know what you are doing differently, I can assure you they are targeting `.Net Core 3.1`

<img width="952" alt="commonnetcore2" src="https://user-images.githubusercontent.com/307872/125878524-1435047a-acff-4127-9b53-5d523fa15efa.PNG">

The builds are done using different virtual machines, so their is no scope for the incorrect version to be included as part of the package. We also use features that aren't present in `.Net 4.0`, so this just isn't possible.

The exception you are seeing is `99%` of the time a result of either missing `VC++` or missing files.

What version of `.Net` are you targeting? What version of the runtime is installed on your target machine?

**Reply by amaitland** [reply]
[IlSpy](https://github.com/icsharpcode/ILSpy/releases/tag/v7.1) confirms `.Net Core 3.1`.

<img width="939" alt="commonnetcore4" src="https://user-images.githubusercontent.com/307872/125881089-33dd6721-dce9-4ff1-a406-35e070166e9b.PNG">

Please confirm you have tested with https://www.nuget.org/api/v2/package/CefSharp.Common.NETCore/91.1.211 (manually download and unzip).

The https://www.nuget.org/api/v2/package/CefSharp.Common/91.1.211 package will have a similar structure, it will include `.Net 4.5.2` dlls in the `Net Core 3.0` folder. This was a hack used until full `.Net Core 3.1` support could be added. This older package would require `.Net 4.5.2` to run the `BrowserSubProcess` that's included by default. The newer `CefSharp.Common.NETCore` package only requires `.Net Core 3.1` or higher.

**Reply by amaitland** [reply]
Using a `Win8.1 IE11` virtual machine

- Removing `.Net 4.5` 
- Installing `Visual C++ 2019 x64`
- Installing `.Net Core 3.1 Desktop Runtime x64`
- Publish [CefSharp.MinimalExample.WinForms.netcore](https://github.com/cefsharp/CefSharp.MinimalExample) (`.Net Core 3.1` and `win-x64`).

The example runs as expected.

I can find no indication the `NetCore` packages are incorrectly targeting `.Net 4`
